src/lib/reasoners/uf.ml

```diff
@@ -999,100 +999,206 @@ let assign_next env =
   Debug.check_invariants "assign_next" env;
   res, env
 
+let save_cache () =
+  LX.save_cache ()
+
+let reinit_cache () =
+  LX.reinit_cache ()
+
 (**** Counter examples functions ****)
-let is_a_good_model_value (x, _) =
-  match X.leaves x with
+let is_a_good_model_value r =
+  match X.leaves r with
     [] -> true
-  | [y] -> X.equal x y
+  | [r'] -> X.equal r r'
   | _ -> false
 
-let is_const_term (x, _) =
-  match X.term_extract x with
+let is_const_term r =
+  match X.term_extract r with
   | Some t, _ ->
     E.const_term t
   | _ ->
-    (*cannot test for theories which don't implement term_extract*)
+    (* Cannot test for theories which don't implement [term_extract]. *)
     true
 
 let model_repr_of_term t env mrepr unbounded =
   try ME.find t mrepr, mrepr
   with Not_found ->
     let mk = try ME.find t env.make with Not_found -> assert false in
     let rep,_ = try MapX.find mk env.repr with Not_found -> assert false in
-    match unbounded with
-    | Some string_repr -> (rep, string_repr), mrepr
-    | None ->
-      let cls =
-        try SE.elements (MapX.find rep env.classes)
-        with Not_found -> assert false
-      in
-      let cls =
-        try List.rev_map (fun s -> s, ME.find s env.make) cls
-        with Not_found -> assert false
-      in
-      let e = X.choose_adequate_model t rep cls in
-      e, ME.add t e mrepr
+    let (rep, hs), mrepr =
+      match unbounded with
+      | Some string_repr ->
+        (rep, Hstring.make string_repr), mrepr
+      | None ->
+        let cls =
+          try SE.elements (MapX.find rep env.classes)
+          with Not_found -> assert false
+        in
+        let cls =
+          try List.rev_map (fun s -> s, ME.find s env.make) cls
+          with Not_found -> assert false
+        in
+        let rep, string_repr = X.choose_adequate_model t rep cls in
+        let hs = Hstring.make string_repr in
+        (rep, hs), ME.add t (rep, hs) mrepr
+    in
+    assert (is_a_good_model_value rep && is_const_term rep);
+    (rep, hs), mrepr
 
 let is_optimization_op = function
   | Sy.Op Sy.Optimize _ -> true
   | _ -> false
 
-let save_cache () =
-  LX.save_cache ()
+(* Caches used during the computation of the model.
+   All these caches are flushed before generating a new model. *)
+module Cache = struct
```

Comment:
Is there a reason why the caches are stored in global variables? It would be cleaner (and thread-safe) to have a type that bundles the hash tables and is passed to `compute_concrete_model_of_val`.

Reply:
No reason, I will create the cache in `compute_concrete_model` directly.